### PR TITLE
Fixed #1205 - 2.0: LiveQuery

### LINKED ISSUE #1205: 2.0: LiveQuery

### PR DESCRIPTION
**Pull request by hideki**
- Implemented LiveQuery
- Added Unit tests

### CODE REVIEW

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    private boolean willUpdate;
+    private boolean forceReload;
+    private long lastUpdatedAt;
+
+    //---------------------------------------------
+    // Constructors
+    //---------------------------------------------
+
+    /* package */ LiveQuery(Query query) {
+        this.query = query;
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: I think we should copy the query.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+     */
+    public void stop() {
+        if (observing) {
+            observing = false;
+            query.getDatabase().removeChangeListener(this);
+        }
+        willUpdate = false; // cancels the delayed update started by -databaseChanged
+    }
+
+    public void addLiveQueryChangeListener(LiveQueryChangeListener listener) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Should be addChangeListener.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        queryChangeListener.add(listener);
+        if (queryChangeListener.size() == 1)
+            run();
+    }
+
+    public void removeLiveQueryChangeListener(LiveQueryChangeListener listener) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Should be removeChangeListener.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java (hunk 1 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: A copy of Query instance will be created in the `ToLive()` method.

**Comment by hideki** [comment]
Applied the feedbacks. Please review this again. Thanks!

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    // Package level access
+    //---------------------------------------------
+
+    /**
+     * The current query results; this updates as the database changes.
+     * Its value will be null until the initial asynchronous query finishes.
+     * <p>
+     * NOTE: for unit test
+     */
+    /*package*/ ResultSet getResultSet() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: I don't see this one get used on LiveQuery class. Can this be removed?

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    //---------------------------------------------
+    // API - public methods
+    //---------------------------------------------
+
+    /**
+     * Starts observing database changes. The .rows property will now update automatically.
+     * (You usually don't need to call this yourself, since accessing or observing the getRows()
+     * property will call -start for you.)
+     */
+    public void run() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Calling run again should force to rerun the query.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            observing = false;
+            query.getDatabase().removeChangeListener(this);
+        }
+        willUpdate = false; // cancels the delayed update started by -databaseChanged
+    }
+
+    public void addChangeListener(LiveQueryChangeListener listener) {
+        queryChangeListener.add(listener);
+        if (queryChangeListener.size() == 1)
+            run();
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: This is not part of the spec. What do you think about this in general?

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    public void removeChangeListener(LiveQueryChangeListener listener) {
+        queryChangeListener.remove(listener);
+        if (queryChangeListener.size() == 0)
+            stop();
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: This is also not part of the spec. Should we include this in spec?

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    private void sendNotification(LiveQueryChange change) {
+        synchronized (queryChangeListener) {
+            for (LiveQueryChangeListener listener : queryChangeListener) {
+                listener.changed(change);
+            }
+        }
+    }
+}
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Should also remove the listener to the database when the object is collected as well.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java (hunk 6 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: This implementation does not work well with the behavior required of `run()` method.

If we add this, we should change `run()` method as a non-public method.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java (hunk 7 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: same with above

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java (hunk 4 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: This is used from unit test only. I will remove this.

**Comment by hideki** [comment]
Please review this PR again. Thanks!

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        this.lastUpdatedAt = 0L;
+    }
+
+    //---------------------------------------------
+    // API - public methods
+    //---------------------------------------------
+
+    /**
+     * Starts observing database changes. The .rows property will now update automatically.
+     * NOTE: Calling run method will (re)start the live query.
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Please update the comment.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    public void run() {
+        observing = true;
+        releaseResultSet();
+        query.getDatabase().addChangeListener(this);
+        update();
+    }
+
+    /**
+     * Stops observing database changes. Calling start() or getRows() will restart it.
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Need update.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        // TODO: Make this asynchronous (as in 1.x)
+
+        if (!observing)
+            return;
+
+        try {
+            Log.i(TAG, "%s: Querying...", this);
+            ResultSet oldResultSet = resultSet;
+            ResultSet newResultSet;
+            if (oldResultSet == null || !oldResultSet.isValidEnumerator())
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: What is the case that the oldResultSet will be invalid? I'm wondering if the enumerator is still valid, if the app iterates to all results?

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+    @Override
+    protected void finalize() throws Throwable {
+        stop();
+        super.finalize();
+    }
+
+    //---------------------------------------------
+    // Package level access
+    //---------------------------------------------
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: Do you prefer to keep this comment when there are no any package level items?

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+            if (newResultSet != null) {
+                if (oldResultSet != null)
+                    Log.i(TAG, "%s: Changed!", this);
+                resultSet = newResultSet;
+                sendNotification(new LiveQueryChange(this, resultSet, null));
+            } else {
+                Log.i(TAG, "%s: ...no change", this);
+            }
+        } catch (CouchbaseLiteException e) {
+            sendNotification(new LiveQueryChange(null, e));
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: The first parameter shouldn't be null.

File: shared/src/main/java/com/couchbase/lite/LiveQueryChange.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+    /*package*/ LiveQueryChange(LiveQuery query, ResultSet rows) {
+        this(query, rows, null);
+    }
+
+    /*package*/ LiveQueryChange(LiveQuery query, Throwable error) {
+        this(query, null, error);
+    }
+
+    /*package*/ LiveQueryChange(LiveQuery query, ResultSet rows, Throwable error) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQueryChange.java]: I think having on this one is enough.

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java (hunk 11 above)
Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: C4QueryEnumerator.next() of native code closes enumerator automatically if it reaches the end of the enumerator. So it is why it checks if the enumerator is still valid.

**Comment by hideki** [comment]
Please re-review this PR again. Thanks!

File: shared/src/main/java/com/couchbase/lite/LiveQuery.java (hunk 11 above)
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/LiveQuery.java]: If it closes automatically, it is not correct to check here as it has a side effect.
For example, 
1. User iterates through the ResultSet -> enumerator closes automatically.
2. There is a database change event but the event doesn't cause the query result to be changed.
3. The update() method is called due to step 2. As the enumerator is not valid anymore, the query is re-run and returns the same old result.

**Comment by hideki** [comment]
Thank you for catching the problem. It is reproducible with the unit test. Please re-review

**Pull request merged**